tutorials/math/combinatorics.md

```diff
@@ -1,10 +1,272 @@
+
+Here we will consider the $i^{th}$ bit from the right. Let's say that the $i^{th}$ bit is set in $k$ out of $n$ numbers in some given subset. If $k$ is odd, then $i^{th}$ bit is set in the XOR of all numbers of the subset, otherwise, it is not set.
+
+Hence if there are $m$ numbers out of $n$ with $i^{th}$ bit set, then the contribution of the bit is $Place\,value\,of\,the\,bit$ $*$ $number\,of\,ways\,to\,get\,odd\,k$ 
+
+Thus we can find $\sum_{k = 1}^{k <= m}$ $m \choose k$ for all odd values of $k$, which comes out to $2^{m - 1}$. Furthermore, we can choose the remaining elements in the subset in $2^{n - m}$ ways by similar logic. Hence total ways to get odd values of $k$ are $2^{n - 1}$, which is independent of both $m$ and $k$.
+
+Hence all we need to do is find bits which are set atleast once (by computing OR) and then multiply the final answer with $2^{n - 1}$.
+Time Complexity of the program is $O(n)$ with $O(1)$ Space Complexity.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    int subsetXORSum(vector<int>& nums) {
+        int arrayOR = 0;
+        // Do OR of whole array to obtain bits which are set atleast once
+        for(int num:nums) arrayOR |= num;
+        // Compute the final answer using the formula discussed
+        return arrayOR*(1<<(nums.size() - 1));
+    }
+};
+```
+
+</TabItem>
+</Tabs>
+
+### Example #3 [62 - Unique Paths](https://leetcode.com/problems/unique-paths/)
+
+This question is a good example of how combinatorics problems can be solved using [Dynamic Programming](../../tutorials/dynamic-programming.md) approaches, just by observing the relations involved. 
+
+Here we observe that a cell can be visited only from the square above or from one to the left. Thus, we can write DP to represent the number of ways to reach the square from $(0, 0)$ as $dp_{i, j}$ = $dp_{i - 1, j}$ $+$ $dp_{i, j - 1}$ 
+
+This algorithm can be implemented in $O(mn)$ time complexity and $O(min(n, m))$ space complexity (Implemented below in $O(mn)$ space in favor of readability).
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    int uniquePaths(int m, int n) {
+      // Make a 2-dimensional DP array with state i, j representing the number of ways to reach the cell (i, j)
+      // The problem can be solved using 1-dimensional DP array using space optimisation 
+      // (not implemented in favour of readablity)
+      vector<vector<int>> dp(m, vector<int> (n, 0));
+      // There is only one way to reach the topmost and leftmost square (base case)
+      dp[0][0] = 1;
+      for(int i = 0; i < m; i++){
+          for(int j = 0; j < n; j++){
+              // Transition: dp[i][j] = dp[i - 1][j] + dp[i][j - 1] 
+              if(i > 0) dp[i][j] += dp[i - 1][j];
+              if(j > 0) dp[i][j] += dp[i][j - 1];
+          }
+      }
+      return dp[m - 1][n - 1];
+    }
+};
+```
```

Comment:
This is DP which is kind of out of topic. What I'm expecting is to use combinatorics to solve this problem.

For example, in python we can use factorial

```py
from math import factorial
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        return factorial(m + n - 2) // factorial(n - 1) // factorial(m - 1)
```

or something like below in C++

```cpp
class Solution {
public:
    template <typename T>
    T binomial(int64_t N, int64_t K) {
      if (K < 0 || N < K) return 0;
      T ret = 1;
      for (T i = 1; i <= K; ++i) {
        ret *= N--;
        ret /= i;
      }
      return ret;
    }
    
    int uniquePaths(int m, int n) {
        return binomial<long long>(m + n - 2, min(n - 1, m - 1));
    }
};
```

Reply:
Hey, I included it on purpose as I felt that knowing how to solve these questions through DP is as important as quite often one may make mistake while doing it through PnC, especially folks who are going to be using this tutorial. DP in that situation is quite ironically an easy way out, especially as people doing leetcode know DP more often. Maybe, I can include the solution through combinatorics as Approach#2.